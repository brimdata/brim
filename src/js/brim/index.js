/* @flow */
import type {FieldValue} from "../types/records"
import type {ReturnType} from "../types"
import ast from "./ast"
import compoundField from "./compoundField"
import dateTuple from "./dateTuple"
import entries from "./entries"
import field from "./field"
import flatRecordsBuffer from "./flatRecordsBuffer"
import form from "./form"
import interop from "./interop"
import log from "./log"
import program from "./program"
import randomHash from "./randomHash"
import record from "./record"
import relTime from "./relTime"
import search from "./search"
import space from "./space"
import span from "./span"
import syntax from "./syntax"
import tab from "./brimTab"
import table from "./table"
import time from "./time"
import zeekLogInfo from "./zeekLogInfo"

export type $Field = {
  name: string,
  type: string,
  value: FieldValue,
  stringValue: () => string,
  queryableValue: () => string,
  compound: () => boolean,
  toCompound: () => $CompoundField,
  toDate: () => Date,
  display: () => string,
  guessWidth: () => number
}

export type $CompoundField = ReturnType<typeof compoundField>
export type $Search = ReturnType<typeof search>
export type $Log = ReturnType<typeof log>

export type Ts = {
  ns: number,
  sec: number
}

export type Span = [Ts, Ts]

export default {
  table,
  program,
  field,
  compoundField,
  log,
  ast,
  syntax,
  space,
  time,
  relTime,
  span,
  dateTuple,
  flatRecordsBuffer,
  form,
  record,
  interop,
  search,
  randomHash,
  entries,
  tab,
  zeekLogInfo
}
