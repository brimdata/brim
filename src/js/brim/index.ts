import ast from "./ast"
import dateTuple from "./dateTuple"
import entries from "./entries"
import form from "./form"
import interop from "./interop"
import program from "./program"
import randomHash from "./randomHash"
import relTime from "./relTime"
import space from "./space"
import span from "./span"
import syntax from "./syntax"
import tab from "./brimTab"

import time from "./time"
import zeekLogInfo from "./zeekLogInfo"
import connection from "./connection"

export type Ts = {
  ns: number
  sec: number
}

export type BrimSpace = ReturnType<typeof space>
export type BrimConnection = ReturnType<typeof connection>

export type Span = [Ts, Ts]

export default {
  program,
  ast,
  syntax,
  space,
  time,
  relTime,
  span,
  dateTuple,
  form,
  interop,
  randomHash,
  entries,
  tab,
  zeekLogInfo,
  connection
}
