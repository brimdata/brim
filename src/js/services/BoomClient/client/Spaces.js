/* @flow */

import type {CreateSpaceBody} from "../types"
import SubClient from "./SubClient"

export default class Spaces extends SubClient {
  list() {
    return this.base.send({
      method: "GET",
      path: "/space"
    })
  }

  get(name: string) {
    return this.base.send({
      method: "GET",
      path: `/space/${encodeURIComponent(name)}`
    })
  }

  create(space: CreateSpaceBody) {
    return this.base.send({
      method: "POST",
      path: "/space",
      payload: space
    })
  }

  delete(name: string) {
    return this.base.send({
      method: "DELETE",
      path: `/space/${encodeURIComponent(name)}`
    })
  }
}
