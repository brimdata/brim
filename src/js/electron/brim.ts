import windowManager, {$WindowManager} from "./tron/windowManager"
import isDev from "./isDev"
import {installExtensions} from "./extensions"
import formatSessionState from "./tron/formatSessionState"
import createGlobalStore from "../state/createGlobalStore"
import {Store} from "redux"
import tron, {Session} from "./tron"
import Prefs from "../state/Prefs"
import {ZQD} from "ppl/zqd/zqd"
import {app} from "electron"
import path from "path"
import {sessionStateFile} from "./tron/session"
import {getGlobalPersistable} from "../state/getPersistable"
import Workspaces from "../state/Workspaces"
import keytar from "keytar"
import os from "os"
import {toAccessTokenKey, toRefreshTokenKey} from "../auth0"

type QuitOpts = {
  saveSession?: boolean
}

type BrimArgs = {
  windows?: $WindowManager
  store?: Store
  session?: Session
  zqd?: ZQD
}

export class Brim {
  readonly windows: $WindowManager
  readonly store: Store
  readonly zqd: ZQD
  readonly session: Session
  public isQuitting = false

  static async boot(
    sessionPath: string = sessionStateFile(),
    createSession = tron.session
  ) {
    const session = createSession(sessionPath)
    const data = await session.load()
    const windows = windowManager(data)
    const store = createGlobalStore(data?.globalState)
    const select = (fn) => fn(store.getState())
    const suricataRunner = select(Prefs.getSuricataRunner)
    const suricataUpdater = select(Prefs.getSuricataUpdater)
    const zeekRunner = select(Prefs.getZeekRunner)
    const spaceDir = path.join(app.getPath("userData"), "data", "spaces")
    const zqd = new ZQD(spaceDir, suricataRunner, suricataUpdater, zeekRunner)
    return new Brim({session, windows, store, zqd})
  }

  constructor(args: BrimArgs = {}) {
    this.windows = args.windows || windowManager()
    this.store = args.store || createGlobalStore(undefined)
    this.session = args.session || tron.session()
    this.zqd = args.zqd || new ZQD(null, null, null, null)
  }

  async start() {
    this.zqd.start()
    if (this.isDev()) await installExtensions()
    this.windows.init()
  }

  activate(winId?: string) {
    if (this.windows.count() === 0) this.windows.init(winId)
  }

  async resetState() {
    // clear keys from secrets storage
    Workspaces.all(this.store.getState()).forEach((ws) => {
      if (ws.authType !== "auth0") return
      keytar.deletePassword(toRefreshTokenKey(ws.id), os.userInfo().username)
      keytar.deletePassword(toAccessTokenKey(ws.id), os.userInfo().username)
    })
    await this.session.delete()
    app.relaunch()
    this.quit({saveSession: false})
  }

  async saveSession() {
    const windowState = await this.windows.serialize()
    const mainState = getGlobalPersistable(this.store.getState())

    await this.session.save(formatSessionState(windowState, mainState))
  }

  async deleteSession() {
    await this.session.delete()
  }

  async quit(opts: QuitOpts = {saveSession: true}) {
    this.isQuitting = true
    if (await this.windows.confirmQuit()) {
      await this.windows.prepareQuit()
      if (opts.saveSession) {
        await this.saveSession()
      }
      this.windows.quit()
      await this.zqd.close()
      app.quit()
    } else {
      this.isQuitting = false
    }
  }

  isDev() {
    return isDev
  }
}
