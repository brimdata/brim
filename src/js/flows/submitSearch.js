/* @flow */
import type {Thunk} from "../state/types"
import {globalDispatch} from "../state/GlobalContext"
import History from "../state/History"
import Investigation from "../state/Investigation"
import Notice from "../state/Notice"
import Search from "../state/Search"
import SearchBar from "../state/SearchBar"
import Tab from "../state/Tab"
import Tabs from "../state/Tabs"
import Viewer from "../state/Viewer"
import brim from "../brim"
import executeHistogramSearch from "./executeHistogramSearch"
import executeTableSearch from "./executeTableSearch"

export default function submitSearch(
  save: Object = {history: true, investigation: true},
  ts: Date = new Date()
): Thunk {
  return function(dispatch, getState) {
    let time = brim.time(ts)
    let prevArgs = Search.getArgs(getState())
    dispatch(SearchBar.submittingSearchBar(ts))
    dispatch(Tab.computeSpan())
    if (!dispatch(SearchBar.validate())) return Promise.reject()

    const state = getState()

    let record = Search.getRecord(state)
    if (save.history) {
      dispatch(History.push(record, time.toTs()))
    }
    if (save.investigation) {
      globalDispatch(Investigation.push(record, time.toTs()))
    }

    let tabId = Tabs.getActive(state)
    let args = Search.getArgs(state)

    dispatch(Viewer.clear(tabId))
    dispatch(Notice.dismiss())

    switch (args.type) {
      case "analytics":
        return dispatch(executeTableSearch(tabId, args))
      case "zoom":
        return dispatch(executeTableSearch(tabId, args))
      default:
        dispatch(executeHistogramSearch(tabId, args, prevArgs))
        return dispatch(executeTableSearch(tabId, args))
    }
  }
}
