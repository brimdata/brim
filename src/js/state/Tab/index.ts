import flows from "./flows"
import {useTabState} from "./local-state"
import reducer from "./reducer"
import selectors from "./selectors"

export default {
  reducer,
  ...selectors,
  ...flows,
  useState: useTabState
}
