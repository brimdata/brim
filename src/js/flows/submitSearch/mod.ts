import {Thunk} from "../../state/types"
import Notice from "../../state/Notice"
import Search from "../../state/Search"
import SearchBar from "../../state/SearchBar"
import Tab from "../../state/Tab"
import Current from "../../state/Current"
import {encodeSearchParams} from "app/search/utils/search-params"
import brim from "src/js/brim"
import Investigation from "src/js/state/Investigation"

type SaveOpts = {history: boolean; investigation: boolean}

export function submitSearch(
  save: SaveOpts = {history: true, investigation: true},
  ts: Date = new Date()
): Thunk<Promise<void>> {
  return function(dispatch, getState) {
    dispatch(Notice.dismiss())
    const record = Search.getCurrentRecord(getState())
    const search = encodeSearchParams(record)
    const workspaceId = Current.getWorkspaceId()
    const spaceId = Current.getSpaceId()

    dispatch(SearchBar.submittingSearchBar(ts))
    dispatch(Tab.computeSpan(ts))

    if (!dispatch(SearchBar.validate())) return Promise.reject()

    const url = `/workspaces/${workspaceId}/lakes/${spaceId}/search?${search}`

    if (save.investigation) {
      dispatch(
        Investigation.push(workspaceId, spaceId, record, brim.time(ts).toTs())
      )
    }
    save.history ? global.tabHistory.push(url) : global.tabHistory.replace(url)
  }
}
