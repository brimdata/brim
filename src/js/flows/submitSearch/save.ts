import {SearchRecord} from "../../types"
import {Thunk} from "../../state/types"
import History from "../../state/History"
import Investigation from "../../state/Investigation"
import brim from "../../brim"

export type SaveOpts = {history: boolean; investigation: boolean}

export function saveToHistory(
  workspaceId: string | null,
  spaceId: string | null,
  record: SearchRecord,
  opts: SaveOpts,
  ts: Date
): Thunk {
  return (dispatch, _) => {
    if (opts.history) {
      dispatch(History.push(record, brim.time(ts).toTs()))
    }
    if (opts.investigation) {
      dispatch(
        Investigation.push(workspaceId, spaceId, record, brim.time(ts).toTs())
      )
    }
  }
}
