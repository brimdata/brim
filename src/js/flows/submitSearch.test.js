/* @flow */

import {createZealotMock} from "zealot"
import Search from "../state/Search"
import SearchBar from "../state/SearchBar"
import Spaces from "../state/Spaces"
import Tab from "../state/Tab"
import brim from "../brim"
import initTestStore from "../test/initTestStore"
import submitSearch from "./submitSearch"

let store, zealot
beforeEach(() => {
  zealot = createZealotMock()
  store = initTestStore(zealot)
  zealot.stubStream("search", [{type: "TaskStart"}, {type: "TaskEnd"}])
})

const spaceInfo = {
  id: "ranch-id",
  name: "ranch-logs",
  span: {
    ts: {sec: 1425564900, ns: 0},
    dur: {sec: 3352893, ns: 750000000}
  },
  pcap_support: true
}

const initTimeWindow = () => (dispatch: Function, getState: Function) => {
  let space = Tab.space(getState())
  if (space) {
    let {min_time, max_time} = space
    dispatch(Search.setSpanArgs([min_time, max_time]))
  }
}

test("fetching a regular search", async () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    initTimeWindow(),
    SearchBar.changeSearchBarInput("_path=conn")
  ])

  store.clearActions()
  await store.dispatch(submitSearch())

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("not saving a search to history", async () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    SearchBar.changeSearchBarInput("_path=conn")
  ])

  store.clearActions()
  await store.dispatch(submitSearch({history: false, investigation: false}))

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an analytic search", async () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    initTimeWindow(),
    SearchBar.changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  await store.dispatch(submitSearch())

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an analytic search without history", async () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    initTimeWindow(),
    SearchBar.changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  await store.dispatch(submitSearch({history: false, investigation: false}))

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an zoom search", async () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    initTimeWindow(),
    Search.setSpanFocus(brim.time.convertToSpan([new Date(0), new Date(1)])),
    SearchBar.changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  await store.dispatch(submitSearch())

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an zoom search without history", async () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    initTimeWindow(),
    SearchBar.changeSearchBarInput("_path=conn | count()"),
    submitSearch()
  ])

  store.clearActions()
  await store.dispatch(submitSearch({history: false, investigation: false}))

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("a bad search query", () => {
  store.dispatchAll([
    Search.setCluster("1"),
    Spaces.setDetail("1", spaceInfo),
    Search.setSpace("ranch-id"),
    SearchBar.changeSearchBarInput("_ath=")
  ])

  store.dispatch(submitSearch()).catch((e) => e)

  expect(store.getActions().map((action) => action.type)).toEqual(
    expect.arrayContaining(["SEARCH_BAR_PARSE_ERROR"])
  )
})
