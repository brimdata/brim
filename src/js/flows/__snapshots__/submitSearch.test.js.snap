// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetching a regular search 1`] = `
Array [
  "SEARCH_BAR_SUBMIT",
  "SEARCH_SPAN_SET",
  "HISTORY_PUSH",
  "VIEWER_CLEAR",
  "NOTICE_DISMISS",
  "CHART_CLEAR",
  "CHART_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "CHART_STATUS",
  "VIEWER_STATUS",
  "CHART_STATUS",
  "HANDLERS_REMOVE",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
]
`;

exports[`fetching an analytic search 1`] = `
Array [
  "SEARCH_BAR_SUBMIT",
  "SEARCH_SPAN_SET",
  "HISTORY_PUSH",
  "VIEWER_CLEAR",
  "NOTICE_DISMISS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "VIEWER_STATUS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
]
`;

exports[`fetching an analytic search without history 1`] = `
Array [
  "SEARCH_BAR_SUBMIT",
  "SEARCH_SPAN_SET",
  "VIEWER_CLEAR",
  "NOTICE_DISMISS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "VIEWER_STATUS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
]
`;

exports[`fetching an zoom search 1`] = `
Array [
  "SEARCH_BAR_SUBMIT",
  "SEARCH_SPAN_SET",
  "HISTORY_PUSH",
  "VIEWER_CLEAR",
  "NOTICE_DISMISS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "VIEWER_STATUS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
]
`;

exports[`fetching an zoom search without history 1`] = `
Array [
  "SEARCH_BAR_SUBMIT",
  "SEARCH_SPAN_SET",
  "VIEWER_CLEAR",
  "NOTICE_DISMISS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "VIEWER_STATUS",
  "VIEWER_STATUS",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
]
`;

exports[`not saving a search to history 1`] = `
Array [
  "SEARCH_BAR_SUBMIT",
  "SEARCH_SPAN_SET",
  "VIEWER_CLEAR",
  "NOTICE_DISMISS",
  "CHART_CLEAR",
  "CHART_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_ABORT",
  "HANDLERS_REGISTER",
  "CHART_STATUS",
  "VIEWER_STATUS",
  "CHART_STATUS",
  "HANDLERS_REMOVE",
  "VIEWER_STATUS",
  "VIEWER_END_STATUS",
  "HANDLERS_REMOVE",
]
`;
